### DISCUSSION TITLE
Migrating pimcore 10.6 version to 11.0, composer udpate got success, error in migration

### DISCUSSION THREAD

**Discussion by Thejaswini-Rao-U** [pimcore/pimcore#15890] [Q&A / Support] [answered]
### Pimcore version

pimcore v 11.0

### Steps to reproduce

When i execute bin/console doctrine:migrations:migrate , migration command getting issue

### Actual Behavior

At first i have updated composer as per document and have resolved issues while updating..

after composer update is successfull, i executed **bin/console doctrine:migrations:migrate** command getting some class error

 i have no idea how to resolve it? 

as i see in the document, its mentioned to rebuild the class after this migration, so i have not executed class rebuild command,

![migrationError](https://github.com/pimcore/pimcore/assets/78895325/53592f21-4f91-4d35-9c20-5ee9555e8b80)

Anyone have idea , please let me know for the solution

Thanks in advance.

### Expected Behavior

success

**Comment by blankse** [comment]
Than you haven't execute the migration in Pimcore 10.6
This migration should save all ClassDefinitions and should make the maxItems to null or int. Before Pimcore 10.4.4 this attributes can be a emty string: https://github.com/pimcore/pimcore/commit/6245efe3be3d64ca5f3369a8472533f98d73043e

Otherwise you can patch manually this attributes in your definiton files (var/classes or config/pimcore/classes dirs)

See also: https://github.com/pimcore/pimcore/discussions/15857

**Comment by Thejaswini-Rao-U** [comment]
@blankse , Thanks for the quick response.

when i executed migration in 10.6 version, it showed me nothing to migrate. 

anyway i will try to do manually and fix.

**Reply by lohitpereira** [reply, marked as answer]
Hi @Thejaswini-Rao-U could you help me with migration to 10.6 at https://github.com/pimcore/pimcore/discussions/16116#discussion-5749234

**Reply by Thejaswini-Rao-U** [reply]
@lohitpereira please check https://github.com/pimcore/pimcore/discussions/16116#discussioncomment-7322559

**Comment by Thejaswini-Rao-U** [comment]
@blankse , how to run specific patch migration command for the classes to migrate in pimcore 11?
because in the 10.6 version after executing migration command, it didnt migrate the classes properly

can you tell, what are the files or datatypes of classes i need to fix it manually,

if fixing manually not good idea, is there any patch migration command execute at this stage?

please suggest?

**Comment by Thejaswini-Rao-U** [comment]
@blankse 
whenever  i execute migrate command in 10.6 version, i get successfull saying , its already in the latest version

please refer below screenshot
![10successmigrate](https://github.com/pimcore/pimcore/assets/78895325/83576e7f-a6a2-4887-8261-b1a24a019864)

it means 10.6 migration is successfull, im getting error while migrating in pimcore 11, 

@blankse  or anyone please tell how to solve it in pimcore 11 without manually doing it ?

**Comment by blankse** [comment]
I see, that the Version20230720161000 migration only save all classes and custom layouts. Maybe there are object bricks and field collections with wrong types. Can you manually save all classes / object bricks / field collections in the pimcore 10.6 admin interface before you migrate to 11?

**Comment by Thejaswini-Rao-U** [comment]
@blankse , thanks for the response

okay i will manually save all the classes/ bricks/collections.

after saving should i run migrate command of 10.6 => ./bin/console doctrine:migrations:migrate --prefix=Pimcore\\Bundle\\CoreBundle
? then update to pimcore 11 ?

is it correct? in this way will this issue resolves?

**Comment by blankse** [comment]
Yes I think so. Your are on the latest 10.6.9 version?

**Comment by Thejaswini-Rao-U** [comment]
yes.

**Comment by Thejaswini-Rao-U** [comment]
@blankse 

after saving classes / object bricks / field collections in the pimcore 10.6 admin interface, /bin/console doctrine:migrations:migrate --prefix=Pimcore\Bundle\CoreBundle, then updated to pimcore 11 version

now after composer update, and while migrating using bin/console doctrine:migrations:migrate command getting different issue related to class definiation

please refer below screenshot

![classbuttonerror1](https://github.com/pimcore/pimcore/assets/78895325/5f3b5f1e-f5d1-4e74-9296-8011ccdacf79)


anyone please suggest the cause of this error and its solution

Thanks in advance

**Comment by blankse** [comment]
@Thejaswini-Rao-U You need to remove all `Button` components from your definitions in 10.6.
It was deprecated in 10.6 and removed in 11. See:
https://github.com/pimcore/pimcore/issues/15123
https://github.com/pimcore/pimcore/issues/15124

**Comment by Thejaswini-Rao-U** [comment]
@blankse , Thanks for the quick response

but what is the alternative solution for this if i remove?

how to implement it if the existing project has button features?

**Comment by Thejaswini-Rao-U** [comment]
@blankse , how to implement button layout custom ? can i get the reference for this with js and how to call it in classes?

is there any other layout which i can use instead button ?

please let me know

**Reply by blankse** [reply]
@Thejaswini-Rao-U I didn't use this Button layout component. I didn't create a custom layout component. So I can't help you there. Maybe @NiklasBr or @fashxp have a Gist for this?

**Comment by Thejaswini-Rao-U** [comment]
@blankse Thanks

@NiklasBr or @fashxp can you share any example reference for button layout ?

**Reply by NiklasBr** [reply]
Can't help, sorry.

**Reply by Thejaswini-Rao-U** [reply]
anyone who knows? please let me know the reference ?

should i implement same as in pimcore 10.6 version as custom ? or it in different way?

**Comment by Thejaswini-Rao-U** [comment]
Hello Team,
if we have to use custom button, then these features were there in 10.6, samething should implement in 11, then what is the use of removing it and making it custom?

**Reply by fashxp** [reply]
The button was removed due to potential security risks. If you need it, you can take the source code from pimcore 10 and put it into a bundle. 
See also the [docs](https://pimcore.com/docs/platform/Pimcore/Extending_Pimcore/Bundle_Developers_Guide/Adding_Object_%20Layout_types) for reference how to add custom layout types.

**Reply by Thejaswini-Rao-U** [reply]
@fashxp , Thanks for the response

The github links which are given in the documents are not opening, getting error page, those files are not exists in the version link given,
none of the links are working from the documents given.

![errorPageGit](https://github.com/pimcore/pimcore/assets/78895325/80ce9127-1c0f-462c-afae-96a33a7be3bf)
![links](https://github.com/pimcore/pimcore/assets/78895325/1f4d3978-4fbe-4adf-86fe-b27c683eb28f)


Documents are not clear, where to add js and php files if created custom? how to call it and how to use it ?

i have created button.php file inside the path var/classes/Layout/ is it right way?

js file where to add ? we should add js files in the bundle where we needed ?

php file how to call in the class or in the class admin UI?

**Reply by blankse** [reply]
@Thejaswini-Rao-U I created a PR to fix the broken links https://github.com/pimcore/pimcore/pull/15907
Is that a help to you?

**Reply by Thejaswini-Rao-U** [reply]
@blankse , https://github.com/pimcore/pimcore/blob/1cba47b3ac4baa2e1f25aaaae0f714e08ded0305/doc/20_Extending_Pimcore/13_Bundle_Developers_Guide/11_Adding_Object_Datatypes.md , document which you have shared link of PR has different extending class ie Create a PHP class for server-side implementation: This class needs to extend Pimcore\Model\DataObject\ClassDefinition\Data and defines how your data type is stored into database

and pimcore 11 version document saying https://pimcore.com/docs/platform/Pimcore/Extending_Pimcore/Bundle_Developers_Guide/Adding_Object_%20Layout_types/ to extend 1) Create a PHP class for server-side implementation: This class needs to extend Pimcore\Model\DataObject\ClassDefinition\Layout

which is correct? its confusing

**Reply by blankse** [reply]
@Thejaswini-Rao-U "Object data type" != "Object layout type"
The button is a layout type, so you need the second link:
https://pimcore.com/docs/platform/Pimcore/Extending_Pimcore/Bundle_Developers_Guide/Adding_Object_%20Layout_types/

**Reply by Thejaswini-Rao-U** [reply]
@blankse  above PR #15907 link doesn't explain how to implement button layout in 11 version then?

**Comment by Thejaswini-Rao-U** [comment]
Any response ?

**Comment by Thejaswini-Rao-U** [comment]
git hub links from the documents are not working for any of the lower versions and even for 11.0
https://github.com/pimcore/pimcore/tree/10.6/bundles/AdminBundle/public/js/pimcore/object/layout

i have added button.php file in the path src/Model/DataObject/ClassDefinition/Layout/Button.php and js file added in the path src/Resources/public/object/classes/layout/button.js, 


have configured 
config/config.yaml

pimcore:
    objects:
        class_definitions:
            layout:
                map:
                  button: \App\Model\DataObject\ClassDefinition\Layout\Button
                  
after this what is the next step, how to call and use it in the object i dont have any idea about this?

Can anyone give the solution to solve this issue?

if we dont get support or proper document for the existing issue then how to go for 11.0 version update?

**Comment by Thejaswini-Rao-U** [comment]
Anyone have idea on this issue? 
![web_toprint](https://github.com/pimcore/pimcore/assets/78895325/ddd6686a-00f9-4817-92a7-6b21b1975d0a)

**Reply by kingjia90** [reply]
Looks like this step could be missing https://github.com/pimcore/pimcore/blob/4dc3b3e4280fb5cc92293224bff19c56ccd21444/doc/23_Installation_and_Upgrade/07_Updating_Pimcore/12_V10_to_V11.md?plain=1#L45-L69

**Comment by Thejaswini-Rao-U** [comment]
@kingjia90 Thanks for the response

i have added above configuration in the file var/config/system_settings/system_settings.yaml, before starting the migration

**Comment by Thejaswini-Rao-U** [comment]
getting below error on migrating

![objectIdError](https://github.com/pimcore/pimcore/assets/78895325/a718fcb8-616d-495a-bc69-6b2498185848)

as i see this is of database table field error, how to fix these on version 11

**Reply by Thejaswini-Rao-U** [reply]
Any response?

**Reply by kingjia90** [reply]
It looks like that the migration already applied the change, from pimcore side it should be idempotent (bug/todo), to fix your local situation, it could help to mark this migration as done or so

**Reply by Thejaswini-Rao-U** [reply]
@kingjia90 , Thanks for the response
how to fix this? lts database table field error

**Reply by kingjia90** [reply]
should be something like `bin/console doctrine:migrations:version 'namespace\Version2022xxxxx' --add` but replace the `namespace\Version2022xxxxx` part with the full migration class name

**Reply by Thejaswini-Rao-U** [reply]
can you specify share the full command to execute for the migration, 

currently i have succesfully completed composer update, now while migrating this command bin/console doctrine:migrations:migrate, got the table field error

**Reply by kingjia90** [reply]
this one, 
![image](https://github.com/pimcore/pimcore/assets/6014195/3f224f58-5c72-4243-800c-01fa396c2d16)
just too lazy to type it down again, though you could select and copy/paste it

**Reply by Thejaswini-Rao-U** [reply]
@kingjia90, Thanks for the response
table filled issue got resolved. thank you so much.

after fixing this issue, when i executed bin/console doctrine:migrations:migrate , got version count error
is it related to db error or version issue?
any specific command need to execute for this?
![color_FIleextension](https://github.com/pimcore/pimcore/assets/78895325/da6f013e-d985-4251-8315-85b37115dcb9)

**Reply by Thejaswini-Rao-U** [reply]
process manager error, using version  "elements/process-manager-bundle": "^5.0.0",
![configurationProcessError](https://github.com/pimcore/pimcore/assets/78895325/5b93bd74-87f5-4883-939f-abc689a0aa10)

**Reply by kingjia90** [reply]
It's a bug on the Elements Process Manager bundle end, sorry can't help.
Please report this problem to their repository and wait for a fix/pr to patch

The property is string only but the setter can be mixed. 
https://github.com/elements-at/ProcessManager/blob/57150433760b6f49bac181813372c9f240ff66ab/src/Model/MonitoringItem.php#L59
https://github.com/elements-at/ProcessManager/blob/57150433760b6f49bac181813372c9f240ff66ab/src/Model/MonitoringItem.php#L541

Looks also like some of the problem we are facing with unserialiazed data and not-nullable-anymore properties eg. https://github.com/pimcore/pimcore/pull/15577 
FYI @ctippler

**Reply by ctippler** [reply]
Thx for the hint. Please try the processManger Version v5.0.3 - there the issue should be gone.

Regards,
Christian

**Reply by Thejaswini-Rao-U** [reply]
@ctippler Thanks for the response

I have installed processManger Version v5.0.3, and then executed bin/console doctrine:migrations:migrate, command , getting below error of process manager itself

NOTE:  previous error was gone , showing different error 
![monitor1Error](https://github.com/pimcore/pimcore/assets/78895325/0b5666b3-1031-406a-9434-affe1d77e737)

**Reply by ctippler** [reply]
Hi, could you please try v5.0.4 :-)

**Reply by Thejaswini-Rao-U** [reply]
I have installed 5.0.4, but still same error exists.

**Reply by Thejaswini-Rao-U** [reply]
Any suggestion to solve this issue??

**Reply by ctippler** [reply]
My Problem is that i can't reproduce it. Normaly the string should be converted here 
![image](https://github.com/pimcore/pimcore/assets/8791970/49dd6c3f-cad2-4ac5-9d61-bc2e91919022)
So maybe you can debug it down.

**Reply by Thejaswini-Rao-U** [reply]
@ctippler, if i found something , i cannot able to change in the plugin bundle right?

is there any other version where this issue will be solved?

which version do you use? you didnt face this issue before?

**Reply by Thejaswini-Rao-U** [reply]
@ctippler , 
in my file also has same code, then why still error exists?
![processmonitoring](https://github.com/pimcore/pimcore/assets/78895325/c254fa3f-d7cb-4370-9adb-09b7cbb7e4ec)

**Comment by Thejaswini-Rao-U** [comment]
Im getting error in admin interface

but i have added file in var/config/admin_system_settings/admin_system_settings.yaml

and the code is below

pimcore_admin:
    config_location:
        admin_system_settings:
            write_target:
                type: 'settings-store'
            read_target:
                type: 'settings-store'
    branding:
        login_screen_invert_colors: false
        color_login_screen: ''
        color_admin_interface: ''
        login_screen_custom_image: ''
        
        any configuration is missing from my side?
        
![erroradmin](https://github.com/pimcore/pimcore/assets/78895325/42fa69db-cedf-4696-8b74-8943c7727a3e)

**Reply by Thejaswini-Rao-U** [reply]
anyone please let me know how to solve ?

**Reply by Thejaswini-Rao-U** [reply]
i have noticed  in the vendor folder in some of the files are still exists old version namespaces

in vendor/pimcore/admin-ui-classic-bundle/src/Controller/Admin/LoginController.php file,  using use Pimcore\Bundle\AdminBundle\Event\AdminEvents; namespace but AdminBundle doesnt exists in 11 version, how this works in vendor then? 


if i replace as per document in my existing bundle by Pimcore\Bundle\SimpleBackendSearchBundle\Event\AdminSearchEvents , AdminSearchEvents events doesnt have 

const LOGIN_REDIRECT = 'pimcore.admin.login.redirect'; const so we cant use AdminSearchEvents instead AdminEvents?

**Reply by Thejaswini-Rao-U** [reply]
anyone having idea of this error showing in admin interface?

please let me know, so that i can complete 11 version migration
![versionCountError](https://github.com/pimcore/pimcore/assets/78895325/0f8f5254-61d0-451c-aa50-2092e3fa28c8)

**Reply by Thejaswini-Rao-U** [reply]
because of process manager issue, migration command is incomplete, also getting version count , admin settings error on admin

please let me know to solve this issue, i will be stuck in this, there is no issue or document regarding these issues.

**Reply by Thejaswini-Rao-U** [reply]
anyone, im waiting for the response? Please share the reasons for this issue and solution?

**Reply by Thejaswini-Rao-U** [reply]
please let me know how to solve version count issue?

**Reply by Thejaswini-Rao-U** [reply]
anyone please help me to solve version count error
i m able to solve process manager issue.
without solving this error im not able to proceed further in admin ui

**Comment by Thejaswini-Rao-U** [comment]
please let me know how to solve version count issue?
